README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
An R Interface to OpenLayers
============================

[![Travis-CI Build Status](https://travis-ci.org/crazycapivara/openlayers.svg?branch=master)](https://travis-ci.org/crazycapivara/openlayers) [![Travis-CI Build Status](https://travis-ci.org/crazycapivara/openlayers.svg?branch=develop)](https://travis-ci.org/crazycapivara/openlayers)

[OpenLayers](https://openlayers.org/) is an open-source JavaScript library *making it easy to put a dynamic map in any web page*. The goal of the openlayers R package is to make this functionality available within R via the [htmlwidgets](https://github.com/ramnathv/htmlwidgets) package. Check [NEWS](NEWS.md) file for available functionality.

Installation
------------

You can install openlayers from github with:

``` r
# install.packages("devtools")
devtools::install_github("crazycapivara/openlayers")

# latest version
devtools::install_github("crazycapivara/openlayers", ref = "develop")
```

Examples
--------

Here we go with some basic examples:

``` r
library(openlayers)
#> openlayers 0.5.0 wrapping openlayersjs v4.6.4
```

``` r
ol() %>%
  add_stamen_tiles() %>%
  set_view(9.5, 51.31667, zoom = 10)

## Points
library("geojsonio")

cities <- us_cities[1:5, ]

ol()  %>%
  add_stamen_tiles() %>%
  add_features(cities, style = icon_style(),
               popup = cities$name)

## Polygons
library("sf")

nc <- st_read(system.file("gpkg/nc.gpkg", package = "sf"),
             quiet = TRUE)

ol() %>%
  add_stamen_tiles("watercolor") %>%
  add_stamen_tiles(
    "terrain-labels",
    options = layer_options(max_resolution = 13000)
  ) %>%
  add_features(
    data = nc,
    style = fill_style("yellow") + stroke_style("blue", 1),
    popup = nc$AREA  
  ) %>%
  add_overview_map()
```

Documentation
-------------

A detailed documentation of the package is still under development, but all functions are documented, so that you can use the build in help functionality of R. Furthermore, check the example scripts in [inst/examples/](https://github.com/crazycapivara/openlayers/blob/master/inst/examples) to get an idea about how to use this package.

Code coverage
-------------

``` r
package_coverage()
#> openlayers Coverage: 61.90%
#> R/legend.R: 0.00%
#> R/styles.R: 22.22%
#> R/vector_tiles.R: 50.00%
#> R/methods.R: 60.00%
#> R/wms.R: 66.67%
#> R/options.R: 90.00%
#> R/controls.R: 100.00%
#> R/dependencies.R: 100.00%
#> R/interactions.R: 100.00%
#> R/ol.R: 100.00%
#> R/plugin-layerswitcher.R: 100.00%
#> R/utils.R: 100.00%
```


NEWS.md
# openlayers 0.5.2

* Fixed bug [#18](https://github.com/crazycapivara/openlayers/issues/13): Display properties as overlay

# openlayers 0.5.1

* Fixed bug when `ol` widget is rendered in shiny via `renderUI`: [issue #13](https://github.com/crazycapivara/openlayers/issues/13)

# openlayers 0.5.0

* Added legend support
  - `add_legend`

# openlayers 0.4.9

* Upgrade to `OpenLayers` v4.6.4
* Added support for `minResolution`, `maxResolution` and `renderMode` options in `addGeojson`
  - options are passed via the `options` parameter in the R functions `add_geojson` and `add_features`
* Added support for styling vector tiles by property

# openlayers 0.4.8

* Upgrade to `OpenLayers` v4.5.0

# openlayers 0.4.7

* `add_<format>_vector_tiles` functions (as introduced in v0.4.6) were replaced by the generic function `add_vector_tiles`, which by default guesses the `format` parameter 
* the stroke color in `stroke_style` can now also be a vector of length equal to the number of features

# openlayers 0.4.6

* Added support for vector tiles, supported formats:
  - mvt (mapbox vector tiles): `add_mapbox_vector_tiles`
  - topojson: `add_topojson_vector_tiles`
  - geojson: `add_geojson_vector_tiles`

# openlayers 0.4.5

* Added wms support
  - `add_wms_tiles`
* Added more parameters to `ol_options`
  - `collapsible_attribution`: fixed or collapsible attribution
  - `zoom_control`: show or hide zoom control 
                       
# openlayers 0.4.4

* Renamed `add_vector_data` (will be removed in a future release) to `add_features`
* Fixed bug in drag and drop caused by new popup parameter

# openlayers 0.4.3

* Added popup support

# openlayers 0.4.2

* Switched to `OpenLayers` v4.3.1

# openlayers 0.4.1

* Added `name` and `type` parameters to `layer_options` 
* Added `ol-layerswitcher` plugin
  - `add_layer_switcher`
  - in order to add layers to the menu just set the `name` parameter of the layer
  - set `type` parameter to `"base"` for base layers
  - layers dropped to the map get random names based on the names-generator from _docker_
* Added horizontal and vertical text offset (parameters) to `text_style`
* Added and updated examples
* As hidden option `renderer` parameter can be set to `webgl` in `ol_options` (should not be used when displaying directly in RStudio)

# openlayers 0.4.0

* Fixed bug in select interaction
* Added more parameters to `icon_style`
* Added drag and drop functionality
  - `add_drag_and_drop`
* Changed default marker icon
* Added `make_icon` function to create marker icons from images, so that they can be used in `icon_style`
* Improved shiny integration by passing data back to R:
  - feature properties on select event
  - coordinates on map click event

# openlayers 0.3.0

* Added full screen control
  - `add_full_screen`
* Added `select_interaction` making features selectable
* Added `ol_options` to set map options like
  - `min_zoom`, `max_zoom`
  - `debug` mode
* Added function `add_vector_data` as a shorthand to `add_geojson` using `geojsonio` to parse any data to geojson

# openlayers 0.2.2

* Added missing function docs
* Added functions to add controls to the map
  - `add_scale_line`
  - `add_overview_map`
  - `add_mouse_control`
* Added examples to `inst/examples`

# openlayers 0.2.1

* Added appendix `_style` to all style functions
  - `style <- stroke_style(color = "blue") + fill_style(color = "yellow")`
* Added possibility to pass vectors of length equal to the number of features to some style functions
  - `style <- fill_style(color = c("green", "blue", "red"))`
* Added documentation to all style functions and some other functions
* Added opacity parameter to `add_geojson` function
* Added R tests for style functions

# openlayers 0.2.0

* Added new approach setting up styles:
  - `style <- stroke(color = "blue") + fill(color = "yellow")`
  - `style <- marker()`
* Added `invoke_method` function to store javascript method calls in `ol`-object
* Added support for multiple vector layers and tile layers

# openlayers 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Added functions to add vector layers from geojson
  - `sf` objects can be added via `geojsonio` package
* Added functions to style features
  - markers are added by applying an icon style to the features


R/methods.R
## TODO: obsolete, because stamen tiles can be added via 'add_stamen_tiles'
## only useful, because 'https' layers sometimes do not show up in RStudio
get_stamen_xyz_url <- function(layer = "toner") {
  sprintf("http://{a-c}.tile.stamen.com/%s/{z}/{x}/{y}.png", layer)
}

#' Get cartodb xyz url
#'
#' for available layers please
#' check \href{https://carto.com/location-data-services/basemaps/}{cartodb-basemaps}
#'
#' @param layer cartodb layer name
#'
#' @return cartodb xyz url for given layer
#'
#' @export
get_cartodb_xyz_url <- function(layer = "dark_all") { # nocov start
  sprintf("https://cartodb-basemaps-{a-c}.global.ssl.fastly.net/%s/{z}/{x}/{y}.png", layer)
} # nocov end

#' Cartodb attribution
#'
#' @export
cartodb_attribution <- function() { # nocov start
  c('&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, ',
    '&copy; <a href="https://carto.com/attribution">CARTO</a>')
} # nocov end

add_cartodb_tiles <- function(ol, layer) {
  stop("needs to be implemented, just do it")
}

#' Set the view of the map (geographical center and zoom)
#'
#' @param ol map widget
#' @param lon longitude of center
#' @param lat latitude of center
#' @param zoom zoom level
#'
#' @export
set_view <- function(ol, lon = 9.5, lat = 51.31667, zoom = 4) {
  invoke_method(ol, "setView", lon, lat, zoom)
}

#' Add tile layers to the map
#'
#' @name add_tiles
#'
#' @param ol map widget
NULL

#' @describeIn add_tiles Add osm tile layer to the map
#'
#' @export
add_osm_tiles <- function(ol, options = layer_options()) {
  invoke_method(ol, "addOSMTiles", camel_case_keys(options))
}

#' @describeIn add_tiles Add stamen tile layer to the map
#'
#' @param layer stamen layer name
#'
#' @export
add_stamen_tiles <- function(ol, layer = "watercolor", options = layer_options()) {
  invoke_method(ol, "addStamenTiles", layer, camel_case_keys(options))
}

#' @describeIn add_tiles Add custom tile layer to the map
#'
#' @param xyz_url xyz url
#' @param attribution attribution (character vector)
#' @param options layer options
#'
#' @examples \dontrun{
#'   xyz_url <- get_cartodb_xyz_url()
#'   attribution <- cartodb_attribution()
#'
#'   ol() %>%
#'     add_xyz_tiles(xyz_url, attribution)
#' }
#'
#' @export
add_xyz_tiles <- function(ol, xyz_url, attribution = NULL, options = layer_options()) {
  invoke_method(ol, "addXYZTiles", xyz_url, attribution, camel_case_keys(options))
}

# TODO: useful ? export : remove
add_geojson_ds <- function(ol, url) {
  invoke_method(ol, "addGeojsonFromUrl", url)
}

#' Add vector layer to the map
#'
#' @param ol map widget
#' @param data geojson, ignored if \code{filename} is given
#' @param filename filename or url to read geojson from
#' @param style style object or \code{NULL} (use default style)
#' @param popup popup texts
#' @param options general layer options, see \code{\link{layer_options}}
#'
#' @examples \dontrun{
#'   require("geojsonio")
#'
#'   geojson <- geojson_json(us_cities[1:10, ])
#'
#'   ol() %>%
#'     add_geojson(geojson)
#' }
#'
#' @export
add_geojson <- function(ol, data = NULL, filename = NULL, style = NULL,
                        popup = NULL, options = layer_options()) {
  if(!is.null(filename)) {
    data <- readr::read_file(filename)
  }
  invoke_method(ol, "addGeojson", data, style, popup, camel_case_keys(options))
}

require_namespace <- function(name) {
  if(!requireNamespace(name, quietly = TRUE)) {
    stop(sprintf("please install %s\n", name), call. = FALSE)
  }
}

#' Add vector layer to the map
#'
#' This function is just a shorthand to \code{\link{add_geojson}} using
#' \pkg{geojsonio} to convert any data object to geojson.
#' Therefore, if the library/namespace is not available, an error is thrown.
#'
#' @param ol map widget
#' @param data any data object which can be parsed to geojson by
#'   \code{\link[geojsonio]{geojson_json}}
#' @param ... optional parameters passed to \code{\link{add_geojson}}
#'
#' @name add_features
#'
#' @export
add_features <- function(ol, data, ...) {
  require_namespace("geojsonio")
  data <- geojsonio::geojson_json(data)
  add_geojson(ol, data, ...)
}

#' @rdname add_features
#'
#' @export
add_vector_data <- function(ol, data, ...) { # nocov start
  .Deprecated("add_features")
  add_features(ol, data, ...)
} # nocov end
